File: packages/weather/src/tradewinds/weather/_iem.py
"""IEM METAR CSV parser — parse comma-separated files from Iowa Environmental Mesonet.

IEM provides pre-parsed METAR fields in US/METAR-native units (°F, kt, mi, inHg).
Emits observation dicts matching specs/observation.json with source="iem".
"""

from __future__ import annotations

import csv
import math
import re
from datetime import datetime
from pathlib import Path
from typing import Any

from tradewinds._internal._bounds import (
    MAX_RAW_METAR_LEN,
    MAX_VISIBILITY_MILES,
    MAX_WX_CODES_LEN,
    MAX_YEAR,
    MIN_YEAR,
    SKY_BASE_MAX_FT,
    SLP_MAX_MB,
    SLP_MIN_MB,
    STATION_CODE_RE,
    TEMP_MAX_C,
    TEMP_MIN_C,
    WIND_DIR_BOUNDS,
    WIND_GUST_MAX,
    WIND_SPEED_MAX,
    bounded_float,
    bounded_float_min,
    bounded_int,
)
from tradewinds._internal._convert import fahrenheit_to_celsius
from tradewinds.weather._awc import icao_to_station_code, map_cloud_cover

_TS_RE = re.compile(r"^(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2})$")


def _safe_float(val: str) -> float | None:
    """Parse IEM string to float. 'M' or empty → None."""
    if not val or val == "M":
        return None
    try:
        f = float(val)
        return f if math.isfinite(f) else None
    except (ValueError, OverflowError):
        return None


def _safe_int(val: str) -> int | None:
    """Parse IEM string to int via float. 'M' or empty → None."""
    f = _safe_float(val)
    return round(f) if f is not None else None


def _parse_precip(val: str) -> float | None:
    """Parse IEM precipitation. 'T' → 0.0, 'M' → None, numeric passthrough."""
    if not val or val == "M":
        return None
    if val.strip().upper() == "T":
        return 0.0
    return _safe_float(val)


def _parse_timestamp(val: str) -> str | None:
    """Parse IEM timestamp to RFC3339.

    '2025-01-01 00:51' → '2025-01-01T00:51:00Z'
    """
    if not val or val == "M":
        return None
    m = _TS_RE.match(val.strip())
    if not m:
        return None
    try:
        datetime.strptime(val.strip(), "%Y-%m-%d %H:%M")
    except ValueError:
        return None
    year = int(m.group(1))
    if not (MIN_YEAR <= year <= MAX_YEAR):
        return None
    return f"{m.group(1)}-{m.group(2)}-{m.group(3)}T{m.group(4)}:{m.group(5)}:00Z"


def _parse_peak_wind_time(val: str) -> str | None:
    """Parse IEM peak_wind_time to HHMM format.

    '2025-01-01 01:31' → '0131'
    """
    if not val or val == "M":
        return None
    val = val.strip()
    parts = val.split(" ")
    if len(parts) != 2:
        return None
    time_parts = parts[1].split(":")
    if len(time_parts) != 2:
        return None
    try:
        h = int(time_parts[0])
        m = int(time_parts[1])
        if 0 <= h <= 23 and 0 <= m <= 59:
            return f"{h:02d}{m:02d}"
    except ValueError:
        pass
    return None


_VALID_OBS_TYPES = frozenset({"METAR", "SPECI"})


def _detect_obs_type(metar: str) -> str:
    """Detect METAR vs SPECI from raw METAR text first word."""
    if not metar or metar == "M":
        return "METAR"
    words = metar.strip().split(None, 1)
    if words and words[0] == "SPECI":
        return "SPECI"
    return "METAR"


def iem_to_observation(
    row: dict[str, str],
    observation_type_override: str | None = None,
) -> dict[str, Any] | None:
    """Convert a single IEM CSV row to an observation schema dict.

    Takes a dict from csv.DictReader (comma-delimited).
    If observation_type_override is set, uses that instead of detecting
    from raw METAR text. IEM strips the SPECI keyword from raw METAR,
    so callers that request report_type=3 or report_type=4 separately
    should pass the override.
    Returns observation dict with exactly 30 fields, or None if row should
    be skipped (invalid station, bad timestamp, or all key vars missing).
    """
    # Station code
    station_raw = row.get("station", "")
    if not station_raw or station_raw == "M":
        return None
    station_code = icao_to_station_code(station_raw)
    if not STATION_CODE_RE.match(station_code):
        return None

    # Timestamp
    observed_at = _parse_timestamp(row.get("valid", ""))
    if observed_at is None:
        return None

    # Observation type: caller override or detect from raw text
    if (
        observation_type_override is not None
        and observation_type_override not in _VALID_OBS_TYPES
    ):
        raise ValueError(
            f"Invalid observation_type_override: {observation_type_override!r}. "
            f"Must be one of {_VALID_OBS_TYPES}"
        )
    metar_text = row.get("metar", "")
    observation_type = observation_type_override or _detect_obs_type(metar_text)

    # Temperature (IEM gives °F, convert to °C — no rounding, then bound)
    raw_temp_f = _safe_float(row.get("tmpf", ""))
    raw_dewp_f = _safe_float(row.get("dwpf", ""))
    temp_c = bounded_float(
        fahrenheit_to_celsius(raw_temp_f), TEMP_MIN_C, TEMP_MAX_C, field="temp_c"
    )
    dewp_c = bounded_float(
        fahrenheit_to_celsius(raw_dewp_f), TEMP_MIN_C, TEMP_MAX_C, field="dewpoint_c"
    )
    # Consistency: if derived °C is out of bounds, the raw °F is also bogus
    temp_f = raw_temp_f if temp_c is not None else None
    dewp_f = raw_dewp_f if dewp_c is not None else None

    # Wind (already in knots)
    wind_dir = bounded_int(_safe_int(row.get("drct", "")), *WIND_DIR_BOUNDS)
    wind_speed = bounded_int(_safe_int(row.get("sknt", "")), 0, WIND_SPEED_MAX)
    wind_gust = bounded_int(_safe_int(row.get("gust", "")), 0, WIND_GUST_MAX)

    # Pressure (already in native units)
    altim = _safe_float(row.get("alti", ""))  # inHg
    slp = _safe_float(row.get("mslp", ""))  # mb/hPa
    if slp is not None and not (SLP_MIN_MB <= slp <= SLP_MAX_MB):
        slp = None

    # Visibility (already in statute miles)
    vis = _safe_float(row.get("vsby", ""))
    if vis is not None:
        if vis < 0:
            vis = None
        else:
            vis = min(vis, MAX_VISIBILITY_MILES)

    # Sky cover and base heights (IEM gives feet, direct passthrough)
    sky_covers: list[str | None] = []
    sky_bases: list[int | None] = []
    for i in range(1, 5):
        cover_raw = row.get(f"skyc{i}", "")
        base_raw = row.get(f"skyl{i}", "")
        cover = map_cloud_cover(cover_raw) if cover_raw and cover_raw != "M" else None
        base = bounded_int(_safe_int(base_raw), 0, SKY_BASE_MAX_FT)
        sky_covers.append(cover)
        sky_bases.append(base)

    # Weather codes (direct passthrough)
    wx_raw = row.get("wxcodes", "")
    weather_codes: str | None = None
    if wx_raw and wx_raw != "M":
        weather_codes = wx_raw[:MAX_WX_CODES_LEN]

    # Precipitation ('T' = trace → 0.0)
    precip = bounded_float_min(_parse_precip(row.get("p01i", "")), 0.0)

    # Snow depth (already in inches)
    snow = bounded_float_min(_safe_float(row.get("snowdepth", "")), 0.0)

    # Peak wind
    pk_gust = bounded_int(_safe_int(row.get("peak_wind_gust", "")), 0, WIND_GUST_MAX)
    pk_dir = bounded_int(_safe_int(row.get("peak_wind_drct", "")), *WIND_DIR_BOUNDS)
    pk_time = _parse_peak_wind_time(row.get("peak_wind_time", ""))

    # Raw METAR (passthrough, truncate)
    raw_metar: str | None = None
    if metar_text and metar_text != "M":
        raw_metar = metar_text[:MAX_RAW_METAR_LEN]

    # All key vars missing → skip row (check raw values for skip decision)
    if all(v is None for v in (raw_temp_f, raw_dewp_f, wind_speed, slp)):
        return None

    return {
        "station_code": station_code,
        "observed_at": observed_at,
        "observation_type": observation_type,
        "source": "iem",
        "temp_c": temp_c,
        "dewpoint_c": dewp_c,
        "temp_f": temp_f,
        "dewpoint_f": dewp_f,
        "wind_dir_degrees": wind_dir,
        "wind_speed_kt": wind_speed,
        "wind_gust_kt": wind_gust,
        "altimeter_inhg": altim,
        "sea_level_pressure_mb": slp,
        "sky_cover_1": sky_covers[0],
        "sky_base_1_ft": sky_bases[0],
        "sky_cover_2": sky_covers[1],
        "sky_base_2_ft": sky_bases[1],
        "sky_cover_3": sky_covers[2],
        "sky_base_3_ft": sky_bases[2],
        "sky_cover_4": sky_covers[3],
        "sky_base_4_ft": sky_bases[3],
        "visibility_miles": vis,
        "weather_codes": weather_codes,
        "precip_1hr_inches": precip,
        "peak_wind_gust_kt": pk_gust,
        "peak_wind_dir": pk_dir,
        "peak_wind_time": pk_time,
        "snow_depth_inches": snow,
        "qc_field": None,
        "raw_metar": raw_metar,
    }


def parse_iem_file(
    path: Path,
    observation_type_override: str | None = None,
) -> list[dict[str, Any]]:
    """Read an IEM CSV file and return list of valid observation dicts.

    Skips comment lines (starting with #) and rows that fail validation.
    If observation_type_override is set, all rows get that type instead
    of detecting from raw METAR text.
    """
    observations: list[dict[str, Any]] = []
    with open(path, newline="", encoding="utf-8", errors="replace") as f:
        filtered = (line for line in f if not line.startswith("#"))
        reader = csv.DictReader(filtered)
        for row in reader:
            obs = iem_to_observation(
                row, observation_type_override=observation_type_override
            )
            if obs is not None:
                observations.append(obs)
    return observations

File: packages-ts/weather/src/_parsers/iem.ts
// IEM METAR CSV parser — string-body in, Observation rows out.
//
// Byte-faithful TS port of Python
// `packages/weather/src/tradewinds/weather/_iem.py::iem_to_observation`
// + `parse_iem_file`. The TS port consumes the CSV body in-memory (a
// string returned by `downloadIemAsos`); the Python version walks a file
// handle but the semantics are identical.
//
// IEM provides pre-parsed METAR fields in US/METAR-native units (°F, kt,
// mi, inHg). We emit Observation dicts matching the canonical
// `schema.observation.v1` shape (30 fields, source="iem").
//
// Key port details:
//  - Comment lines starting with `#` are stripped BEFORE the CSV header
//    is consumed (mirrors Python's `filtered = (line for line in f if
//    not line.startswith("#"))`).
//  - `M` and empty string both parse as `null` for numeric fields.
//  - `T` (trace) parses as `0.0` for `p01i` only; on other numeric
//    fields, T is non-numeric and yields `null`.
//  - Out-of-bounds Celsius consistency rule: when derived `temp_c` is
//    nulled by bounds, the raw `temp_f` is ALSO nulled (Python L170-171).
//  - All 4 key vars (raw `tmpf`, raw `dwpf`, `wind_speed`, `slp`) missing
//    → row skipped (returns `null` from `iemToObservation`).
//  - Output dict-key order matches Python verbatim so `JSON.stringify`
//    produces byte-stable output across SDKs (downstream diff tooling).
//
// CSV implementation: hand-rolled split-on-comma. IEM's `format=comma`
// endpoint does NOT quote fields with embedded commas (the Python parser
// confirms this empirically — it uses csv.DictReader with default
// dialect and works). A hand-rolled split keeps the bundle-size gate
// happy (no papaparse dep) and matches the TS Architect rubric §2
// (bundle size) + the plan's "DO NOT add deps" note.

import {
  MAX_RAW_METAR_LEN,
  MAX_VISIBILITY_MILES,
  MAX_WX_CODES_LEN,
  MAX_YEAR,
  MIN_YEAR,
  SKY_BASE_MAX_FT,
  SLP_MAX_MB,
  SLP_MIN_MB,
  STATION_CODE_RE,
  TEMP_MAX_C,
  TEMP_MIN_C,
  WIND_DIR_BOUNDS,
  WIND_GUST_MAX,
  WIND_SPEED_MAX,
  boundedFloat,
  boundedFloatMin,
  boundedInt,
} from "@tradewinds/core/internal/bounds";
import { fahrenheitToCelsius } from "@tradewinds/core/internal/convert";

import { type Observation, icaoToStationCode, mapCloudCover } from "./awc.js";

const [WIND_DIR_LO, WIND_DIR_HI] = WIND_DIR_BOUNDS;

// Match Python `_TS_RE = re.compile(r"^(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2})$")`.
const TS_RE = /^(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2})$/;

const VALID_OBS_TYPES = new Set<string>(["METAR", "SPECI"]);

export type IemObservationTypeOverride = "METAR" | "SPECI";

export interface IemToObservationOptions {
  /**
   * Force `observation_type` to this value regardless of metar text.
   * Used by callers that issue separate report_type=3/4 fetches and
   * therefore know the type per-batch without inspecting the raw text.
   */
  observationTypeOverride?: IemObservationTypeOverride;
}

/** Raw IEM CSV row (header → cell value), all strings before parsing. */
export type IemCsvRow = Record<string, string>;

/**
 * Parse an IEM numeric cell into a finite float. `"M"` (IEM's missing
 * sentinel) and empty strings yield `null`; non-finite values yield `null`.
 *
 * Mirrors Python `_safe_float`.
 */
function safeFloat(val: string): number | null {
  if (val === "" || val === "M") return null;
  const f = Number(val);
  return Number.isFinite(f) ? f : null;
}

/**
 * Parse an IEM numeric cell into a rounded integer via `safeFloat`.
 * `null` propagates. Mirrors Python `_safe_int` (round → int).
 */
function safeInt(val: string): number | null {
  const f = safeFloat(val);
  return f === null ? null : Math.round(f);
}

/**
 * Parse the IEM precipitation cell. `M` / empty → `null`; `T` (trace) → `0.0`;
 * numeric values pass through `safeFloat`. Mirrors Python `_parse_precip`.
 */
function parsePrecip(val: string): number | null {
  if (val === "" || val === "M") return null;
  if (val.trim().toUpperCase() === "T") return 0.0;
  return safeFloat(val);
}

/**
 * Parse IEM timestamp (`YYYY-MM-DD HH:MM`) into RFC3339 / ISO 8601 UTC.
 * Returns `null` for missing (`""`/`M`), malformed, calendar-invalid, or
 * out-of-year-range inputs. Mirrors Python `_parse_timestamp`.
 */
function parseTimestamp(val: string): string | null {
  if (val === "" || val === "M") return null;
  const trimmed = val.trim();
  const m = TS_RE.exec(trimmed);
  if (m === null) return null;
  // RegExp matched — groups 1..5 are guaranteed defined.
  const year = Number.parseInt(m[1] as string, 10);
  const month = Number.parseInt(m[2] as string, 10);
  const day = Number.parseInt(m[3] as string, 10);
  const hour = Number.parseInt(m[4] as string, 10);
  const minute = Number.parseInt(m[5] as string, 10);
  // Range checks BEFORE Date.UTC round-trip (rejects Feb-30 / hour-25 inputs
  // that Date.UTC would otherwise silently roll forward).
  if (year < MIN_YEAR || year > MAX_YEAR) return null;
  if (month < 1 || month > 12) return null;
  if (day < 1 || day > 31) return null;
  if (hour > 23 || minute > 59) return null;
  // Calendar validity via UTC round-trip — rejects "2025-02-30 12:00" etc.
  const millis = Date.UTC(year, month - 1, day, hour, minute, 0, 0);
  if (!Number.isFinite(millis)) return null;
  const d = new Date(millis);
  if (
    d.getUTCFullYear() !== year ||
    d.getUTCMonth() !== month - 1 ||
    d.getUTCDate() !== day ||
    d.getUTCHours() !== hour ||
    d.getUTCMinutes() !== minute
  ) {
    return null;
  }
  // Format as `YYYY-MM-DDTHH:MM:00Z` byte-faithful with Python's f-string.
  return `${m[1]}-${m[2]}-${m[3]}T${m[4]}:${m[5]}:00Z`;
}

/**
 * Parse `YYYY-MM-DD HH:MM` peak-wind-time into the bare `HHMM` form Python
 * emits. Returns `null` on `""`/`M`/malformed. Mirrors Python
 * `_parse_peak_wind_time`.
 */
function parsePeakWindTime(val: string): string | null {
  if (val === "" || val === "M") return null;
  const trimmed = val.trim();
  const parts = trimmed.split(" ");
  if (parts.length !== 2) return null;
  const timeParts = (parts[1] as string).split(":");
  if (timeParts.length !== 2) return null;
  const h = Number.parseInt(timeParts[0] as string, 10);
  const min = Number.parseInt(timeParts[1] as string, 10);
  if (!Number.isFinite(h) || !Number.isFinite(min)) return null;
  if (h < 0 || h > 23 || min < 0 || min > 59) return null;
  const hh = h < 10 ? `0${h}` : String(h);
  const mm = min < 10 ? `0${min}` : String(min);
  return `${hh}${mm}`;
}

/**
 * Detect METAR vs SPECI from the raw METAR text first word. Empty / `M` →
 * `"METAR"` (safe default). Mirrors Python `_detect_obs_type`.
 */
function detectObsType(metar: string): "METAR" | "SPECI" {
  if (metar === "" || metar === "M") return "METAR";
  const words = metar.trim().split(/\s+/, 1);
  if (words.length > 0 && words[0] === "SPECI") return "SPECI";
  return "METAR";
}

/**
 * Convert one IEM CSV row (header→cell dict) into the canonical
 * Observation row. Returns `null` if the row should be skipped:
 *  - station missing/M, or doesn't match STATION_CODE_RE after K-strip
 *  - timestamp unparseable / out-of-range
 *  - ALL 4 key vars (raw tmpf, raw dwpf, wind_speed, slp) missing
 *
 * Throws on bad `observationTypeOverride` (Python L152-156 parity).
 *
 * Output dict-key order is preserved verbatim — matters for downstream
 * byte-stable JSON snapshot diffs.
 */
export function iemToObservation(
  row: IemCsvRow,
  opts: IemToObservationOptions = {},
): Observation | null {
  // --- Station code ------------------------------------------------------
  const stationRaw = row.station ?? "";
  if (stationRaw === "" || stationRaw === "M") return null;
  const stationCode = icaoToStationCode(stationRaw);
  if (!STATION_CODE_RE.test(stationCode)) return null;

  // --- Timestamp ---------------------------------------------------------
  const observedAt = parseTimestamp(row.valid ?? "");
  if (observedAt === null) return null;

  // --- Observation type (caller override or auto-detect) ----------------
  const override = opts.observationTypeOverride;
  if (override !== undefined && !VALID_OBS_TYPES.has(override)) {
    throw new Error(
      `Invalid observation_type_override: ${JSON.stringify(
        override,
      )}. Must be one of ${[...VALID_OBS_TYPES].join(", ")}`,
    );
  }
  const metarText = row.metar ?? "";
  const observationType: "METAR" | "SPECI" =
    override !== undefined ? override : detectObsType(metarText);

  // --- Temperature (IEM gives °F; derive °C; bounds-check on °C) ---------
  const rawTempF = safeFloat(row.tmpf ?? "");
  const rawDewpF = safeFloat(row.dwpf ?? "");
  const tempC = boundedFloat(fahrenheitToCelsius(rawTempF), TEMP_MIN_C, TEMP_MAX_C, {
    field: "temp_c",
  });
  const dewpC = boundedFloat(fahrenheitToCelsius(rawDewpF), TEMP_MIN_C, TEMP_MAX_C, {
    field: "dewpoint_c",
  });
  // Consistency rule (Python L170-171): if derived °C is out of bounds,
  // the raw °F is also bogus — null both sides.
  const tempF: number | null = tempC !== null ? rawTempF : null;
  const dewpF: number | null = dewpC !== null ? rawDewpF : null;

  // --- Wind (already in knots) ------------------------------------------
  const windDir = boundedInt(safeInt(row.drct ?? ""), WIND_DIR_LO, WIND_DIR_HI);
  const windSpeed = boundedInt(safeInt(row.sknt ?? ""), 0, WIND_SPEED_MAX);
  const windGust = boundedInt(safeInt(row.gust ?? ""), 0, WIND_GUST_MAX);

  // --- Pressure ----------------------------------------------------------
  const altim = safeFloat(row.alti ?? ""); // inHg (passthrough)
  let slp = safeFloat(row.mslp ?? ""); // mb / hPa
  if (slp !== null && !(slp >= SLP_MIN_MB && slp <= SLP_MAX_MB)) {
    slp = null;
  }

  // --- Visibility (statute miles; clamp at MAX_VISIBILITY_MILES) --------
  let vis = safeFloat(row.vsby ?? "");
  if (vis !== null) {
    if (vis < 0) vis = null;
    else if (vis > MAX_VISIBILITY_MILES) vis = MAX_VISIBILITY_MILES;
  }

  // --- Sky covers + base heights (4 columns; IEM gives feet) ------------
  const skyCovers: Array<string | null> = [];
  const skyBases: Array<number | null> = [];
  for (let i = 1; i <= 4; i += 1) {
    const coverRaw = row[`skyc${i}`] ?? "";
    const baseRaw = row[`skyl${i}`] ?? "";
    const cover = coverRaw !== "" && coverRaw !== "M" ? mapCloudCover(coverRaw) : null;
    const base = boundedInt(safeInt(baseRaw), 0, SKY_BASE_MAX_FT);
    skyCovers.push(cover);
    skyBases.push(base);
  }

  // --- Weather codes (truncate to MAX_WX_CODES_LEN) ---------------------
  const wxRaw = row.wxcodes ?? "";
  const weatherCodes: string | null =
    wxRaw !== "" && wxRaw !== "M" ? wxRaw.slice(0, MAX_WX_CODES_LEN) : null;

  // --- Precipitation ('T' = trace → 0.0; clamp at 0.0) ------------------
  const precip = boundedFloatMin(parsePrecip(row.p01i ?? ""), 0.0);

  // --- Snow depth (clamp at 0.0) ----------------------------------------
  const snow = boundedFloatMin(safeFloat(row.snowdepth ?? ""), 0.0);

  // --- Peak wind ---------------------------------------------------------
  const pkGust = boundedInt(safeInt(row.peak_wind_gust ?? ""), 0, WIND_GUST_MAX);
  const pkDir = boundedInt(safeInt(row.peak_wind_drct ?? ""), WIND_DIR_LO, WIND_DIR_HI);
  const pkTime = parsePeakWindTime(row.peak_wind_time ?? "");

  // --- Raw METAR (truncate) ---------------------------------------------
  const rawMetar: string | null =
    metarText !== "" && metarText !== "M" ? metarText.slice(0, MAX_RAW_METAR_LEN) : null;

  // --- Skip-row gate: ALL 4 key vars missing → drop the row -------------
  // Mirror Python L222-224: uses the RAW values (raw_temp_f, raw_dewp_f,
  // wind_speed bounded, slp bounded). Note slp uses post-bounds null so a
  // pressure outside [870, 1084] also counts as missing for skip purposes.
  if (rawTempF === null && rawDewpF === null && windSpeed === null && slp === null) {
    return null;
  }

  // Return literal preserves Python's iem_to_observation key order
  // exactly — 30 fields. JSON.stringify ordering matters for downstream
  // byte-stable diff tooling.
  return {
    station_code: stationCode,
    observed_at: observedAt,
    observation_type: observationType,
    source: "iem",
    temp_c: tempC,
    dewpoint_c: dewpC,
    temp_f: tempF,
    dewpoint_f: dewpF,
    wind_dir_degrees: windDir,
    wind_speed_kt: windSpeed,
    wind_gust_kt: windGust,
    altimeter_inhg: altim,
    sea_level_pressure_mb: slp,
    sky_cover_1: skyCovers[0] as string | null,
    sky_base_1_ft: skyBases[0] as number | null,
    sky_cover_2: skyCovers[1] as string | null,
    sky_base_2_ft: skyBases[1] as number | null,
    sky_cover_3: skyCovers[2] as string | null,
    sky_base_3_ft: skyBases[2] as number | null,
    sky_cover_4: skyCovers[3] as string | null,
    sky_base_4_ft: skyBases[3] as number | null,
    visibility_miles: vis,
    weather_codes: weatherCodes,
    precip_1hr_inches: precip,
    peak_wind_gust_kt: pkGust,
    peak_wind_dir: pkDir,
    peak_wind_time: pkTime,
    snow_depth_inches: snow,
    qc_field: null, // IEM CSV doesn't carry a QC field — schema-stable null
    raw_metar: rawMetar,
  };
}

/**
 * Parse a full IEM CSV body into Observation rows.
 *
 * - Lines beginning with `#` are dropped before the header is consumed
 *   (mirror Python `filtered = (line for line in f if not line.startswith("#"))`).
 * - The first non-comment line is the header; subsequent lines are data.
 * - Rows that `iemToObservation` rejects are silently dropped (parser
 *   never throws on bad data; only `observationTypeOverride` validation
 *   throws — that's a programmer-error path).
 *
 * Returns an empty array for empty / header-only / all-comments input.
 */
export function parseIemCsv(
  csvBody: string,
  opts: IemToObservationOptions = {},
): ReadonlyArray<Observation> {
  if (csvBody === "") return [];
  // Normalize line endings + split. Drop empty trailing lines so trailing
  // newlines don't yield phantom rows.
  const lines = csvBody.split(/\r?\n/).filter((line) => !line.startsWith("#"));
  // Skip empty lines (incl. the final newline tail).
  const nonEmpty = lines.filter((line) => line.length > 0);
  if (nonEmpty.length === 0) return [];
  const headerLine = nonEmpty[0] as string;
  const header = headerLine.split(",");
  const out: Observation[] = [];
  for (let i = 1; i < nonEmpty.length; i += 1) {
    const line = nonEmpty[i] as string;
    const cells = line.split(",");
    const row: IemCsvRow = {};
    for (let c = 0; c < header.length; c += 1) {
      const key = header[c] as string;
      row[key] = (cells[c] ?? "").trim();
    }
    const obs = iemToObservation(row, opts);
    if (obs !== null) out.push(obs);
  }
  return out;
}
